ec_hostcmd: add support for serial host interface

barnas-michal opened the pull request:
This PR adds support for SHI (Serial Host Interface) on the ITE and NPCX chips.
It's used to handle the host commands between EC and AP on ARM and RISC-V boards.

barnas-michal committed:
ec_host_cmd: add NPCX SHI peripheral for the host commands

This commit adds the support for host commands being transported
by the Serial Host Interface on the NPCX SoC.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
ec_host_cmd: add ITE SHI peripheral for the host commands

This commit adds the support for host commands being transported
by the Serial Host Interface on the IT8xxx2 SoC.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

fabiobaltieri commented on dts/riscv/ite/it8xxx2.dtsi:
```diff
@@ -860,6 +860,18 @@
+			reg = <0x00f03a00 0x30>;
+			interrupts = <171 0>;
+			interrupt-parent = <&intc>;
+			pinctrl-0 = <&shi_mosi_gpm0_default
+				     &shi_miso_gpm1_default
+				     &shi_clk_gpm4_default
+				     &shi_cs_gpm5_default>;
+			pinctrl-names = "default";
```
I think these should not be in this file and go in the boards using it instead.

semihalf-barnas-michal replied:
Shouldn't the default interfaces be defined in the chip device tree?

fabiobaltieri replied:
The basic device node yes (it's an intrinsic characteristic of the SoC) but the pin mapping is a board  design decision so it should go in the board file.

semihalf-barnas-michal replied:
Had moved this to the EC boards dts, forgot to resolve the conversation.

fabiobaltieri commented on dts/arm/nuvoton/npcx.dtsi:
```diff
@@ -354,6 +354,15 @@
+			reg = <0x4000f000 0x120>;
+			interrupts = <18 1>;
+			clocks = <&pcc NPCX_CLOCK_BUS_APB3 NPCX_PWDWN_CTL5 1>;
+			shi-cs-wui =<&wui_io53>;
```
Add a space after `=`

MulinChao commented on drivers/ec_host_cmd_periph/Kconfig:
```diff
@@ -31,6 +31,55 @@ config EC_HOST_CMD_PERIPH_ESPI
+
+config EC_HOST_CMD_PERIPH_SHI_MAX_REQUEST
+	hex "Max data size for the version 3 request packet"
+	default 0x220 if EC_HOST_CMD_PERIPH_SHI_NPCX
+	default 0x100 if EC_HOST_CMD_PERIPH_SHI_ITE
```
In npcx series, no matter input or output buffer size is 128 bytes. EC_HOST_CMD_PERIPH_SHI_MAX_REQUES is not related to hardware capability directly. But it does influence the performance of flash write command during updating RW firmware. (Please refer [shi.c](https://chromium.googlesource.com/chromiumos/platform/ec/+/refs/heads/main/chip/npcx/shi.c#122) for more detail)

semihalf-barnas-michal replied:
I am not aware of any hardware limitations to this.

semihalf-barnas-michal replied:
Thanks @MulinChao for clarification.
@fabiobaltieri : it can't be a dts property, since this has to be exported to EC code. There is a host command that must send the response size value to the AP. Extracting it from DTS won't be good, since this function can be shared between eSPI and SHI.

fabiobaltieri replied:
I think you could still keep the Kconfig option and change the default based on a dt property, seems like it's mainly used for clock frequency around the code base, for example:  https://github.com/zephyrproject-rtos/zephyr/blob/7c9d7922116c9ce7fded4213f3bbbbbbbbe1c382/soc/arm/atmel_sam0/saml21/Kconfig.defconfig.series#L28-L29

Thinking is if it's a property of the SHI device node, it may be handy to keep all of them together in the dt rather than having one in the driver Kconfig. That said finding the node like is kinda messy. Meh, not sure, up to you :-)

semihalf-barnas-michal replied:
As we talked directly, the Kconfig for the maximum request and response sizes have to be exported outside the kernel to the application, so it must be visible independently of the device tree. Also the Kconfig doesn't allow to set the value of DT property minus some const (padding), so there is no good way to set value basing on it. I will just export the fifo size to DTS, leaving this Kconfig as independent value.

semihalf-barnas-michal commented on drivers/ec_host_cmd_periph/ec_host_cmd_periph_shi_ite.c:
```diff
@@ -0,0 +1,525 @@
+
+BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 1, "Invalid number of ITE SHI peripherals");
+
+#define SHI_NODE DT_NODELABEL(shi0)
```
Done

semihalf-barnas-michal commented on drivers/ec_host_cmd_periph/ec_host_cmd_periph_shi_ite.c:
```diff
@@ -0,0 +1,525 @@
+
+#define SHI_NODE DT_NODELABEL(shi0)
+
+LOG_MODULE_REGISTER(host_cmd_shi_ite, LOG_LEVEL_ERR);
```
Nice idea, thanks!

barnas-michal committed:
ec_host_cmd: add Kconfig to set ec host commands logging level

This commit adds Kconfig CONFIG_EC_HC_LOG_LEVEL that sets the logging
level of host commands subsystem.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

semihalf-barnas-michal commented on drivers/ec_host_cmd_periph/ec_host_cmd_periph_shi_ite.c:
```diff
@@ -0,0 +1,525 @@
+}
+
+/* This routine handles spi received unexcepted data */
+static void spi_bad_received_data(const struct device *dev, int count)
```
The SHI is (using) the SPI with some additional logic, so some names can be confusing. I will change it.

semihalf-barnas-michal commented on drivers/ec_host_cmd_periph/ec_host_cmd_periph_shi_ite.c:
```diff
@@ -0,0 +1,525 @@
+
+static void shi_ite_int_handler(const void *arg)
+{
+	const struct device *dev = DEVICE_DT_GET(DT_DRV_INST(0));
```
> (...as in the npcx file) :-)

Ouch, that hurts. Thanks

semihalf-barnas-michal commented on drivers/ec_host_cmd_periph/ec_host_cmd_periph_shi_ite.c:
```diff
@@ -0,0 +1,525 @@
+
+void shi_ite_cs_callback(const struct device *port, struct gpio_callback *cb, gpio_port_pins_t pins)
+{
+	const struct device *dev = DEVICE_DT_GET(DT_DRV_INST(0));
```
I needed the dev instead of data, which could be taken using container_of, but I changed the function to take the data and used your suggestion.

semihalf-barnas-michal commented on drivers/ec_host_cmd_periph/ec_host_cmd_periph_shi_npcx.c:
```diff
@@ -0,0 +1,998 @@
+
+	switch (action) {
+	case PM_DEVICE_ACTION_SUSPEND:
+		shi_npcx_disable(dev);
```
I understand, these callbacks are for device power management, but they will be enabled/disabled based on the host power state. There's no other way to allow to disable and enable the interface outside the driver code.
Here's how it looks like in the EC code: https://chromium-review.googlesource.com/c/chromiumos/platform/ec/+/3827184

ChiHuaL replied:
Thank you for the explanation of using the PM_DEVICE* power management to disable/enable the SHI.
I understand now how it works with the HOST's power state.
Does it make sense to use  pm_device_runtime_disable(set SHI PM_STATE to RESUME)/pm_device_runtime_enable(set SHI PM_STATE to SUSPEND) instead of pm_device_runtime_get/pm_device_runtime_out to enable/disable SHI?
I asked this because I read the the [pm device mangement document](https://docs.zephyrproject.org/3.1.0/services/pm/device_runtime.html#introduction) and saw:
```
Therefore, applications can not manually suspend or resume a device. An application can, however, decide when to  
disable or enable runtime power management for a device. This can be useful, for example, if an application wants a
particular device to be always active.
```

semihalf-barnas-michal replied:
> Does it make sense to use pm_device_runtime_disable(set SHI PM_STATE to RESUME)/pm_device_runtime_enable(set SHI PM_STATE to SUSPEND) instead of pm_device_runtime_get/pm_device_runtime_out to enable/disable SHI?

Sure, it sounds like I shouldn't use the put/get in the EC application, even if it isn't used inside the HC driver. I will make a change to the code there. It shouldn't require any change in this PR.
Thank you very much for checking the code and mentioning this documentation, I wasn't aware of that.

keith-zephyr replied:
I think you need to add a call to `pm_device_init_suspended()` as part of `shi_npcx_init()` so that when the EC application calls `pm_device_runtime_disable()`, the PM layer will transition the SHI driver from suspended to active.   [PM device implementation guidelines](https://docs.zephyrproject.org/latest/services/pm/device_runtime.html#implementation-guidelines)

semihalf-barnas-michal replied:
Yea, right, otherwise it would fire the callback unnecessarily. But the enable and disable are supposed to only be in the EC application, right?

keith-zephyr replied:
I think the driver init is supposed to call both `pm_device_init_suspended() and `pm_device_runtime_enable()`.  This ensures the internal PM state is initialized correctly.

Example [here](https://github.com/zephyrproject-rtos/zephyr/blob/42db096d284016b9352965ddabe82148e5d614e2/drivers/dai/intel/dmic/dmic.c#L877).  

I find only [instance](https://github.com/zephyrproject-rtos/zephyr/blob/42db096d284016b9352965ddabe82148e5d614e2/drivers/power_domain/power_domain_gpio.c#L99) where `pm_device_runtime_enable()` isn't called - so I assume this is a bug.

semihalf-barnas-michal commented on drivers/ec_host_cmd_periph/ec_host_cmd_periph_shi_ite.c:
```diff
@@ -0,0 +1,525 @@
+
+	/* Configure the SPI chip select */
+	status = gpio_pin_configure(cfg->cs_port, cfg->cs_pin,
+				    GPIO_INPUT | IT8XXX2_GPIO_VOLTAGE_1P8);
```
Done

semihalf-barnas-michal commented on drivers/ec_host_cmd_periph/ec_host_cmd_periph_shi_ite.c:
```diff
@@ -0,0 +1,525 @@
+/* Max data size for a version 3 request/response packet. */
+#define SPI_MAX_REQUEST_SIZE SPI_RX_MAX_FIFO_SIZE
+#define SPI_MAX_RESPONSE_SIZE                                                                      \
+	(SPI_TX_MAX_FIFO_SIZE - EC_SPI_PREAMBLE_LENGTH - EC_SPI_PAST_END_LENGTH)
```
They weren't, removed

GTLin08 commented on drivers/ec_host_cmd_periph/Kconfig:
```diff
@@ -31,6 +35,55 @@ config EC_HOST_CMD_PERIPH_ESPI
+config EC_HOST_CMD_PERIPH_SHI_MAX_RESPONSE
+	hex "Max data size for the version 3 response packet"
+	default 0x220 if EC_HOST_CMD_PERIPH_SHI_NPCX
+	default 0x100 if EC_HOST_CMD_PERIPH_SHI_ITE
```
In ITE SHI, the maximum data size of response should not be greater than 248bytes, which needs to be added with EC_SPI_PREAMBLE_LENGTH + EC_SPI_PAST_END_LENGTH.
SPI_TX_MAX_FIFO_SIZE is not equal to EC_HOST_CMD_PERIPH_SHI_MAX_RESPONSE.

semihalf-barnas-michal replied:
Right, I've merged two defines into one. Thanks for noticing it and correcting me, I will change this.

fabiobaltieri commented on drivers/ec_host_cmd_periph/ec_host_cmd_periph_shi_npcx.c:
```diff
@@ -0,0 +1,992 @@
+
+BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 1, "Invalid number of NPCX SHI peripherals");
+
+LOG_MODULE_REGISTER(host_cmd_shi_npcx, LOG_LEVEL_DBG);
```
Think you meant to use `CONFIG_EC_HC_LOG_LEVEL` here

semihalf-barnas-michal replied:
Right, forgot about this...

fabiobaltieri commented on drivers/ec_host_cmd_periph/ec_host_cmd_periph_shi_npcx.c:
```diff
@@ -0,0 +1,992 @@
+	/* Reset shi's state machine for error recovery */
+	shi_npcx_reset_prepare(dev);
+
+	LOG_DBG("END\n");
```
Drop `\n`, it's implicit with `LOG_` macros.

fabiobaltieri commented on drivers/ec_host_cmd_periph/ec_host_cmd_periph_shi_ite.c:
```diff
@@ -0,0 +1,516 @@
+ *
+ * Some commands can continue for a while. This function is called by
+ * host_command when it completes.
+ *
```
nit: drop empty comment line

fabiobaltieri commented on drivers/ec_host_cmd_periph/ec_host_cmd_periph_shi_ite.c:
```diff
@@ -0,0 +1,516 @@
+	 * indicates reached target count(IT83XX_SPI_FTCB1R,
+	 * IT83XX_SPI_FTCB0R) and the length field of the host
+	 * requested data.
+	 */
```
nit: span this comment block on the 80 columns? Should save a line

semihalf-barnas-michal replied:
Fixed some more comments

fabiobaltieri commented on dts/bindings/shi/nuvoton,npcx-shi.yaml:
```diff
@@ -0,0 +1,47 @@
+      For example the WUI mapping on NPCX7 would be
+        shi-cs-wui = <&wui_io53>;
+
+  buffer-rx-size:
+    type: int
+    required: true
+    description: |
+      Size of the internal buffer used for receiving data on the bus.
+
+  buffer-tx-size:
+    type: int
+    required: true
+    description: |
+      Size of the internal buffer used for transmitting data over the bus.
```
Would it make sense to add these two to a common template here?

semihalf-barnas-michal replied:
Sure it does, done

fabiobaltieri commented on dts/bindings/shi/nuvoton,npcx-shi.yaml:
```diff
@@ -0,0 +1,35 @@
+
+compatible: "nuvoton,npcx-shi"
+
+include: [base.yaml, pinctrl-device.yaml, shi-buffer-size.yaml]
```
I was actually thinking more of doing `shi-device.yaml`, and then that could include `base.yaml`, and then this would only be `[shi-device.yaml, pinctrl-device.yaml]`. Does it make sense?

semihalf-barnas-michal replied:
Oh, sorry. Couldn't make free bindings due to the different method of CS handling, but cleaned definitely.

fabiobaltieri replied:
Cool, one last bit: I think the pinctrl entries should stay on the higher level file, that's because the constraint change depending on the soc (such as having one or two entries), so by having them on the soc specific binding you can have correct "required" properties and assume they are going to be there on the driver.

semihalf-barnas-michal replied:
That sounds right, hope that now it is as you thought

GTLin08 commented on drivers/ec_host_cmd_periph/ec_host_cmd_periph_shi_ite.c:
```diff
@@ -0,0 +1,504 @@
+	}
+
+	/* Enable SPI chip select pin interrupt */
+	status = gpio_pin_interrupt_configure(cfg->cs.port, cfg->cs.pin,
```
The gpio interrupt enable here is duplicated declaration as line 440.
The line 440 declaration should be reserved and the flag set to edge falling is right.

semihalf-barnas-michal replied:
Thanks for catching this, removed the redundant configuration.

GTLin08 commented on drivers/ec_host_cmd_periph/ec_host_cmd_periph_shi_ite.c:
```diff
@@ -0,0 +1,504 @@
+	 * has been completed and AP has received the last byte which is EC_SHI_PAST_END from
+	 * EC responded data, then AP ended the transaction.
+	 */
+	if (IT83XX_SPI_ISR & IT83XX_SPI_ENDDETECTINT) {
```
In this condition, we can use pm_policy_state_lock_put() to make this device enter the power policy to save power. [pm_state=PM_STATE_STANDBY ](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/riscv/riscv-ite/common/power.c#L25)

semihalf-barnas-michal replied:
Added these functions, thank you very much

GTLin08 commented on drivers/ec_host_cmd_periph/ec_host_cmd_periph_shi_ite.c:
```diff
@@ -0,0 +1,504 @@
+void shi_ite_cs_callback(const struct device *port, struct gpio_callback *cb, gpio_port_pins_t pins)
+{
+	struct shi_it8xxx2_data *data = CONTAINER_OF(cb, struct shi_it8xxx2_data, cs_cb);
+
```
When this GPIO interrupt is fired, it means that SHI starts to transmit data. We should prevent the device entering the power policy, we can use `pm_policy_state_lock_get()`.

semihalf-barnas-michal replied:
Added these functions, thank you very much

MulinChao reviewed (dismissed):
No further comments for npcx driver.

keith-zephyr commented on drivers/ec_host_cmd_periph/Kconfig:
```diff
@@ -35,6 +35,45 @@ config EC_HOST_CMD_PERIPH_ESPI
+choice
+	prompt "SHI driver"
+	default EC_HOST_CMD_PERIPH_SHI_NPCX if SOC_FAMILY_NPCX
+
```
It is recommended that `choice` symbols are assigned a name: `choice EC_HOST_CMD_PERIPH_SHI_DRIVER`.

semihalf-barnas-michal replied:
Should the above choice for `prompt "Host commands peripheral"` also have a name?

keith-zephyr replied:
A separate small PR would be appropriate.

semihalf-barnas-michal replied:
Will do!

keith-zephyr commented on drivers/ec_host_cmd_periph/ec_host_cmd_periph_shi_ite.c:
```diff
@@ -0,0 +1,504 @@
+	return 0;
+}
+
+static int shi_ite_init(const struct device *dev)
```
Add call to [pm_device_init_suspended()](https://docs.zephyrproject.org/latest/services/pm/api/index.html#c.pm_device_init_suspended) for compatibility with the EC application.

semihalf-barnas-michal commented on drivers/ec_host_cmd_periph/ec_host_cmd_periph_shi_npcx.c:
```diff
@@ -0,0 +1,983 @@
+LOG_MODULE_REGISTER(host_cmd_shi_npcx, CONFIG_EC_HC_LOG_LEVEL);
+
+/* Driver convenience defines */
+#define HAL_INSTANCE(dev) (struct shi_reg *)((const struct shi_npcx_config *)(dev)->config)
```
Right... I improperly merged the defines, thanks for good catch!

semihalf-barnas-michal commented on drivers/ec_host_cmd_periph/ec_host_cmd_periph_shi.h:
```diff
@@ -0,0 +1,101 @@
+ * request, the AP will clock in bytes until it sees the framing byte, then
+ * clock in the response packet.
+ */
+#define EC_SHI_FRAME_START 0xec
```
Sure, it may be worth talking about best possible way to deduplicate this code between zephyr and EC (legacy cros ec).
